subsys/audio_module/audio_module.c

```diff
@@ -0,0 +1,1065 @@
+		break;
+
+	default:
+		LOG_DBG("Invalid module type %d for module %s", handle->description->type,
+			handle->name);
+		return -EINVAL;
+	}
+
+	sys_slist_init(&handle->hdl_dest_list);
+	k_mutex_init(&handle->dest_mutex);
+
+	handle->thread_id = k_thread_create(
+		&handle->thread_data, handle->thread.stack, handle->thread.stack_size,
+		(k_thread_entry_t)module_thread_in_out, (void *)handle, NULL, NULL,
+		K_PRIO_PREEMPT(handle->thread.priority), 0, K_FOREVER);
+
+	ret = k_thread_name_set(handle->thread_id, &handle->name[0]);
+	if (ret) {
+		LOG_DBG("Failed to start thread for module %s thread, ret %d", handle->name, ret);
+
+		/* Clean up the handle */
+		memset(handle, 0, sizeof(struct audio_module_handle));
+		return ret;
+	}
+
+	handle->state = AMOD_STATE_CONFIGURED;
+
+	k_thread_start(handle->thread_id);
+
+	LOG_DBG("Start thread");
+
+	return 0;
+}
+
+/**
+ * @brief  Function to close an open module.
+ */
+int audio_module_close(struct audio_module_handle *handle)
+{
+	int ret;
+
+	if (handle == NULL) {
+		LOG_DBG("Module handle is NULL");
+		return -EINVAL;
+	}
+
+	if (handle->state == AMOD_STATE_UNDEFINED || handle->state == AMOD_STATE_RUNNING) {
+		LOG_DBG("Module %s in an invalid state, %d, for close", handle->name,
+			handle->state);
+		return -ENOTSUP;
+	}
+
+	if (handle->description->functions->close != NULL) {
+		ret = handle->description->functions->close(
+			(struct audio_module_handle_private *)handle);
+		if (ret) {
+			LOG_DBG("Failed close call to module %s, returned %d", handle->name, ret);
+			return ret;
+		}
+	}
+
+	if (handle->thread.msg_rx != NULL) {
+		data_fifo_empty(handle->thread.msg_rx);
+	}
+
+	if (handle->thread.msg_tx != NULL) {
+		data_fifo_empty(handle->thread.msg_tx);
+	}
+
+	/*
+	 * How to return all the data to the slab items?
+	 * Test the semaphore and wait for it to be zero.
+	 */
+
+	k_thread_abort(handle->thread_id);
+
+	LOG_DBG("Closed module %s", handle->name);
+
+	/* Clean up the handle */
+	memset(handle, 0, sizeof(struct audio_module_handle));
+
+	return 0;
+};
+
+/**
+ * @brief  Function to reconfigure a module.
+ */
+int audio_module_reconfigure(struct audio_module_handle *handle,
+			     struct audio_module_configuration *configuration)
+{
+	int ret;
+
+	if (handle == NULL || configuration == NULL) {
+		LOG_DBG("Module input parameter error");
+		return -EINVAL;
+	}
+
+	if (handle->state == AMOD_STATE_UNDEFINED || handle->state == AMOD_STATE_RUNNING) {
+		LOG_DBG("Module %s in an invalid state, %d, for setting the configuration",
+			handle->name, handle->state);
+		return -ENOTSUP;
+	}
+
+	if (handle->description->functions->configuration_set != NULL) {
+		ret = handle->description->functions->configuration_set(
+			(struct audio_module_handle_private *)handle, configuration);
+		if (ret) {
+			LOG_DBG("Set configuration for module %s send failed, ret %d", handle->name,
+				ret);
+			return ret;
+		}
+	} else {
+		LOG_DBG("Reconfiguration for module %s has no set configuration function",
```

Comment:
If the module has no reconfig function, should we print a WRN and return here? It is a tad strange to set the module to state CONFIGURED on line 612 if it is not "possible".

Reply:
This is an interesting one as some audio modules may get their configuration from the data stream directly. In this case the modules we have all require some form of configuration data so this can be changed to either a WRN or probably better make it an ERR?

Reply:
Have added a check that the configure and process function pointers are not NULL when the module is opened and we added the __ASSERT  to test the process is not NULL.